ambry-replication/src/main/java/com.github.ambry.replication/BlobIdTransformer.java

```diff
@@ -158,10 +158,15 @@ private Message newMessage(InputStream inputStream, StoreKey newKey, MessageInfo
       PutMessageFormatInputStream putMessageFormatInputStream =
           new PutMessageFormatInputStream(newKey, blobEncryptionKey, newProperties, userMetaData, blobData.getStream(),
               blobData.getSize(), blobData.getBlobType());
+      // Reuse the original CRC if present in the oldMessageInfo. This is important to ensure that messages that are
+      // received via replication are sent to the store with proper CRCs (which the store needs to detect duplicate
+      // messages). As an additional guard, here the original CRC is only reused if the key's ID in string form is the
+      // same after conversion.
+      Long originalCrc = oldMessageInfo.getStoreKey().getID().equals(newKey.getID()) ? oldMessageInfo.getCrc() : null;
       MessageInfo info =
           new MessageInfo(newKey, putMessageFormatInputStream.getSize(), false, oldMessageInfo.isTtlUpdated(),
```

Comment:
a test for this (you can put it into some existing ones I think)?

Reply:
Good idea, added.